drivers/serial/uart_bcm2711.c

```diff
@@ -0,0 +1,328 @@
+#define BCM2711_MU_IIR_FLUSH		0xc6
+
+#define BCM2711_MU_LCR_7BIT		0x02
+#define BCM2711_MU_LCR_8BIT		0x03
+
+#define BCM2711_MU_LSR_TX_IDLE		BIT(6)
+#define BCM2711_MU_LSR_TX_EMPTY		BIT(5)
+#define BCM2711_MU_LSR_RX_OVERRUN	BIT(1)
+#define BCM2711_MU_LSR_RX_READY		BIT(0)
+
+#define BCM2711_MU_CNTL_RX_ENABLE	BIT(0)
+#define BCM2711_MU_CNTL_TX_ENABLE	BIT(1)
+
+struct bcm2711_uart_config {
+	DEVICE_MMIO_ROM; /* Must be first */
+	mem_addr_t base;
+	uint32_t baud_rate;
+	uint32_t clocks;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(const struct device *dev);
+#endif
+};
+
+struct bcm2711_uart_data {
+	DEVICE_MMIO_RAM; /* Must be first */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+#endif
+};
+
+bool bcm2711_mu_lowlevel_can_getc(mem_addr_t base)
+{
+	return sys_read32((base + BCM2711_MU_LSR)) & BCM2711_MU_LSR_RX_READY;
+}
+
+uint8_t bcm2711_mu_lowlevel_getc(mem_addr_t base)
+{
+	/* Wait until there is no data in the FIFO */
+	while (!bcm2711_mu_lowlevel_can_getc(base))
+		;
+
+	return (sys_read32((base + BCM2711_MU_IO)) & 0xFF);
+}
+
+bool bcm2711_mu_lowlevel_can_putc(mem_addr_t base)
+{
+	return sys_read32(base + BCM2711_MU_LSR) & BCM2711_MU_LSR_TX_EMPTY;
+}
+
+void bcm2711_mu_lowlevel_putc(mem_addr_t base, uint8_t ch)
+{
+	/* Wait until there is data in the FIFO */
+	while (!bcm2711_mu_lowlevel_can_putc(base))
+		;
+
+	/* Send the character */
+	sys_write32(ch, (base + BCM2711_MU_IO));
+}
+
+void bcm2711_mu_lowlevel_init(mem_addr_t base, bool skip_baudrate_config,
+			      uint32_t baudrate, uint32_t input_clock)
+{
+	uint32_t val = 0;
+	uint32_t divider;
+	mem_addr_t ier = base + BCM2711_MU_IER;
+	mem_addr_t lcr = base + BCM2711_MU_LCR;
+	mem_addr_t baud = base + BCM2711_MU_BAUD;
+	mem_addr_t cntl = base + BCM2711_MU_CNTL;
+
+	/* Wait until there is data in the FIFO */
+	while (!bcm2711_mu_lowlevel_can_putc(base))
+		;
+
+	/* Disable port */
+	sys_write32(0x0, cntl);
+
+	/* Disable interrupts */
+	sys_write32(0x0, ier);
+
+	/* Setup 8bit data width and baudrate */
+	sys_write32(BCM2711_MU_LCR_8BIT, lcr);
+	if (!skip_baudrate_config) {
+		divider = (input_clock / (baudrate * 8));
+		sys_write32(divider - 1, baud);
+	}
+
+	/* Enable RX & TX port */
+	val = BCM2711_MU_CNTL_RX_ENABLE | BCM2711_MU_CNTL_TX_ENABLE;
+	sys_write32(val, cntl);
+}
+
+/**
+ * @brief Initialize UART channel
+ *
+ * This routine is called to reset the chip in a quiescent state.
+ * It is assumed that this function is called only once per UART.
+ *
+ * @param dev UART device struct
+ *
+ * @return 0
+ */
+static int uart_bcm2711_init(const struct device *dev)
+{
+	const struct bcm2711_uart_config *uart_cfg = (dev)->config;
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+	bcm2711_mu_lowlevel_init(DEVICE_MMIO_GET(dev), 1, uart_cfg->baud_rate, uart_cfg->clocks);
+	return 0;
+}
+
+static void uart_bcm2711_poll_out(const struct device *dev, unsigned char c)
+{
+	bcm2711_mu_lowlevel_putc(DEVICE_MMIO_GET(dev), c);
```

Comment:
get rid of `DEVICE_MMIO_GET`

Reply:
I have refactored `uart_bcm2711.c`, and I assure you that you won't see as many annoying `DEVICE_MMIO_GET `and parentheses.